packages/image/src/Image.svelte

```diff
@@ -0,0 +1,198 @@
+<script context="module">
+	import {
+		providers,
+		domains,
+		device_sizes,
+		image_sizes
+	} from 'virtual:__svelte-image-options__.js';
+	import { DEV } from 'esm-env';
+
+	const all_sizes = image_sizes.concat(device_sizes);
+
+	/**
+	 * Calculate the widths to use for the srcset using the following logic:
+	 * - If the sizes prop is set, use the vw sizes in the sizes prop.
+	 * 	 Calculate the smallest ratio to the device size and use that to filter out all sizes that will never be used.
+	 * - Always filter out all sizes above 2x the width (because upscaling beyond 2x is wasteful)
+	 * - Never use widths outside the device/image sizes because many CDNs work with fixed sizes to prevent abuse
+	 *
+	 * @param {number} width
+	 * @param {boolean} fixed
+	 * @param {string | undefined} sizes
+	 */
+	function getWidths(width, fixed, sizes) {
+		if (sizes) {
+			const viewport_width = /(^|\s)(1?\d?\d)vw/g;
+			const percent_sizes = [];
+			for (let match; (match = viewport_width.exec(sizes)); match) {
+				percent_sizes.push(parseInt(match[2]));
+			}
+			if (percent_sizes.length) {
+				const smallest_ratio = Math.min(...percent_sizes) * 0.01;
+				return {
+					widths: all_sizes.filter((s) => s >= device_sizes[0] * smallest_ratio && s <= width * 2),
+					kind: 'w'
+				};
+			}
+			return { widths: all_sizes.filter((s) => s <= width * 2), kind: 'w' };
+		}
+		if (!fixed) {
+			return { widths: device_sizes.filter((s) => s <= width * 2), kind: 'w' };
+		}
+
+		const widths = [
+			...new Set(
+				// > This means that most OLED screens that say they are 3x resolution,
+				// > are actually 3x in the green color, but only 1.5x in the red and
+				// > blue colors. Showing a 3x resolution image in the app vs a 2x
+				// > resolution image will be visually the same, though the 3x image
+				// > takes significantly more data. Even true 3x resolution screens are
+				// > wasteful as the human eye cannot see that level of detail without
+				// > something like a magnifying glass.
+				// https://blog.twitter.com/engineering/en_us/topics/infrastructure/2019/capping-image-fidelity-on-ultra-high-resolution-devices.html
+				[width, width * 2].map(
+					(w) => all_sizes.find((p) => p >= w) || all_sizes[all_sizes.length - 1]
+				)
+			)
+		];
+		return { widths, kind: 'x' };
+	}
+
+	/**
+	 * @param {string} src
+	 */
+	function matches_domain(src) {
+		const url = new URL(src, 'http://n'); // if src is protocol relative, use dummy domain
+		if (url.href === src) {
+			return domains.some((domain) => url.hostname === domain);
+		} else {
+			return true; // relative urls are always ok
+		}
+	}
+</script>
+
+<script>
+	/** @type {string | import('vite-imagetools').Img | import('vite-imagetools').Picture} */
+	export let src;
```

Comment:
do we want to be so tightly coupled to `vite-imagetools`?

```suggestion
	/** @type {string} */
	export let src;
```